guppylang/module.py

```diff
@@ -93,184 +95,129 @@ def load(self, m: Union[ModuleType, "GuppyModule"]) -> None:
+            else:
+                self._raw_defs[defn.id] = defn
+            if instance is not None:
+                self._globals.impls.setdefault(instance.id, {})
+                self._globals.impls[instance.id][defn.name] = defn.id
+            else:
+                self._globals.names[defn.name] = defn.id
 
-    def register_type(self, name: str, defn: TypeDef) -> None:
-        """Registers an existing Guppy type as belonging to this Guppy module."""
-        self._check_not_yet_compiled()
-        self._check_type_name_available(name, None)
-        self._globals.type_defs[name] = defn
+    def register_func_def(
+        self, f: PyFunc, instance: TypeDef | None = None
+    ) -> RawFunctionDef:
+        """Registers a Python function definition as belonging to this Guppy module."""
+        defn = RawFunctionDef(DefId.fresh(self), f.__name__, None, f, get_py_scope(f))
+        self.register_def(defn, instance)
+        return defn
 
-    def register_type_var(self, name: str, linear: bool) -> None:
-        """Registers a new type variable"""
-        self._check_not_yet_compiled()
-        self._check_type_name_available(name, None)
-        self._globals.param_vars[name] = TypeParam(
-            len(self._globals.param_vars), name, linear
-        )
+    def register_func_decl(
+        self, f: PyFunc, instance: TypeDef | None = None
+    ) -> RawFunctionDecl:
+        """Registers a Python function declaration as belonging to this Guppy module."""
+        decl = RawFunctionDecl(DefId.fresh(self), f.__name__, None, f)
+        self.register_def(decl, instance)
+        return decl
 
-    def _register_buffered_instance_funcs(self, defn: TypeDef) -> None:
+    def _register_buffered_instance_funcs(self, instance: TypeDef) -> None:
         assert self._instance_func_buffer is not None
         buffer = self._instance_func_buffer
         self._instance_func_buffer = None
-        for f in buffer.values():
-            if isinstance(f, CustomFunction):
-                self.register_custom_func(f, defn)
-            else:
-                is_def, pyfunc = f
-                if is_def:
-                    self.register_func_def(pyfunc, defn)
-                else:
-                    self.register_func_decl(pyfunc, defn)
+        for defn in buffer.values():
+            self.register_def(defn, instance)
 
     @property
     def compiled(self) -> bool:
         return self._compiled
 
+    @staticmethod
+    def _check_defs(
+        raw_defs: Mapping[DefId, RawDef], globals: Globals
+    ) -> dict[DefId, CheckedDef]:
+        """Helper method to parse and check raw definitions."""
+        raw_globals = globals | Globals(raw_defs, {}, {}, {})
+        parsed = {
+            defn.id: defn.parse(raw_globals) if isinstance(defn, ParsableDef) else defn
+            for defn in raw_defs.values()
+        }
+        parsed_globals = globals | Globals(parsed, {}, {}, {})
+        return {
+            defn.id: (
```

Comment:
It seems a bit of a code smell that we only use the values here - it seems we have a dictionary where all of the values contain their key